libmachine/persist/secretstore.go

```diff
@@ -0,0 +1,210 @@
+package persist
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/rancher/machine/libmachine/host"
+	"github.com/rancher/machine/libmachine/log"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	corev1types "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type secretStore struct {
+	Store
+	SecretName, SecretNamespace string
+	SecretClient                corev1types.SecretInterface
+	secret                      *v1.Secret
+}
+
+func NewSecretStore(store Store, secretName, secretNamespace, kubeConfigPath string) (Store, error) {
+	var config *rest.Config
+	var err error
+	if kubeConfigPath != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &secretStore{Store: store, SecretName: secretName, SecretNamespace: secretNamespace, SecretClient: clientset.CoreV1().Secrets(secretNamespace)}
+	if err := s.extractConfig(); err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func (s *secretStore) Remove(name string) error {
+	if err := s.Store.Remove(name); err != nil {
+		return fmt.Errorf("error removing directories for host %v: %v", name, err)
+	}
+	return s.saveSecret(name)
+}
+
+func (s *secretStore) Save(host *host.Host) error {
+	if err := s.Store.Save(host); err != nil {
+		return fmt.Errorf("error saving with file store: %v", err)
+	}
+	return s.saveSecret(host.Name)
+}
+
+func (s *secretStore) saveSecret(hostName string) error {
+	// create the tar.gz file
+	destFile := &bytes.Buffer{}
+
+	fileWriter := gzip.NewWriter(destFile)
+	tarfileWriter := tar.NewWriter(fileWriter)
+
+	if err := s.addDirToArchive(tarfileWriter); err != nil {
+		tarfileWriter.Close()
+		fileWriter.Close()
+		return err
+	}
+
+	tarfileWriter.Close()
+	fileWriter.Close()
+
+	secret, err := s.loadSecret()
+	if err != nil {
+		return fmt.Errorf("unable to load secret : %v", err)
+	}
+
+	if secret.Data == nil {
+		secret.Data = make(map[string][]byte)
+	}
+
+	if bytes.Compare(secret.Data["extractedConfig"], destFile.Bytes()) == 0 {
+		return nil
+	}
+
+	secret.Data["extractedConfig"] = destFile.Bytes()
```

Comment:
throw extractedConfig key as a constant since you use across 2 diff functions

Reply:
Done.